src/osd/PrimaryLogPG.cc

```diff
@@ -2344,6 +2344,16 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
   maybe_force_recovery();
 }
 
+bool PrimaryLogPG::is_dedup_chunk(const chunk_info_t& chunk)
+{
+  if (pg_pool_t::fingerprint_t fp_algo = pool.info.get_fingerprint_type();
+      chunk.has_reference() &&
+      fp_algo != pg_pool_t::TYPE_FINGERPRINT_NONE) {
```

Comment:
If set-chunk is called without using with-reference, it is possible for the chunk (without the reference) to be in the map.

Reply:
No, it's not.  Your implementation calls inc_refcount_by_set in the event that the set-chunk is called without a reference and waits for the reference to be taken before updating chunk_map.  Also, it would be *incorrect* for a chunk to point at a dedup target it does not hold a refcount for because there'd be no guarantee that the target doesn't get removed, which would mean data loss.

Reply:
I don't understand the point -- you want to say. 
1. If set-chunk (--with-reference) is called, there is a chunk which has the reference.
2. If set-chunk (without --with-reference) is called, there is a chunk which has **no** reference.

The first case above will call inc_refcount_by_set(). In addition, there are chunks which are either deduped
or normal chunked in the manifest object.

You mean dec_refcount_by_dirty() does not need to be called when the manifest object has no reference?

Or you mean the set-chunk should be changed like it only supports --with-reference case (the first case)?

Reply:
Perhaps I have misunderstood.  Let me try to lay out my assumptions.

First, note that if a mapping (offset, len) -> foo exists in object_info.chunk_map on disk in the upper tier, we assume that we are free to fulfil a rados read containing the region [offset, offset + len) from the target object foo.  This implies that foo must exist.  If for some reason foo does not exist, we cannot fulfil the read, and therefore we have data loss.  The mechanism we use to ensure that foo really does exist as long as our mapping does is that we take a reference count for every mapping pointing at object foo.  Thus, I conclude that **every mapping must have a corresponding refcount.** (with some extra complications for chunks common to adjacent clones)

From the above, I inferred from your SET_CHUNK implementation that you intended two possible usages:
1. The user (tiering agent code either local to the osd or remote) wants to redirect an offset, length extent to an object that it has already written out, but not taken a reference to.
2. The user (tiering agent code either local to the osd or remote) wants to redirect an offset, length extent to an object that it has already written out *and has already taken a reference to*.

I assumed that case 1. would use the with-reference variant and case 2. would not.  However, the end result of both case 1 and case 2 is that the object holds a reference to foo, which is why it's safe to place foo into chunk_map and rely on it for future rados reads.

Reply:
Or is there some other rationale here?  Is chunk_map meant to be used for use-cases other than dedup?

Reply:
Ok, I just want to clarify the policy ( **every mapping must have a corresponding refcount** ).

Reply:
From my point of view, the policy in play here is that *data loss is unacceptable*.  This implementation assumes fundamentally that chunk_map contains references to objects that really exist because we rely on them to fulfil user reads.  The mechanism we use to ensure that the target object outlives any references to it is the refcount.  Therefore, an entry in chunk_map that refers to foo without holding a refcount to foo could lead to data loss as foo could be deleted, and is therefore unacceptable.

I'm not trying to impose a specific policy here, I'm trying to reconcile your design and implementation with itself, your stated goals,  and the basic requirements of ceph specifically and a storage system generally.  All I really have to go by are the contents of your PRs and these conversations.  If there's something I'm missing, please let me know.  For instance, are you intending to use chunk_map for things other than refcounted, immutable dedup chunks that have different rules?  Otherwise, this code appears not to fit with the remainder of the design we've discussed for the reasons I've given.

In summary, as far as I can tell, *your design* requires that every mapping hold a valid reference with a refcount.

Reply:
> From my point of view, the policy in play here is that data loss is unacceptable. 
I agree.

Could you take a quick look a commit I added ?